### ISSUE TITLE
Change § Testing and Debugging to use "fact" vs "assertion" terminology.

### ISSUE THREAD

**Issue by cgranade** [MicrosoftDocs/quantum-docs-pr#420]
The current [testing and debugging article](https://docs.microsoft.com/quantum/techniques/testing-and-debugging#logging-and-assertions) uses the term "assertion" to include both assertions (that is, diagnostic operations that depend on state owned by a target machine) and facts (diagnostic functions that depend only on their actual inputs). We should clarify, as these two kinds of diagnostic conditions are semantically distinct, and are referred to in the Q# standard libraries by different names. For example, [`Microsoft.Quantum.Diagnostics.Fact`](https://docs.microsoft.com/qsharp/api/qsharp/microsoft.quantum.diagnostics.fact) is a function, while [`Microsoft.Quantum.Diagnostics.AssertQubitIsInStateWithinTolerance`](https://docs.microsoft.com/qsharp/api/qsharp/microsoft.quantum.diagnostics.assertqubitisinstatewithintolerance) uses the verb "assert" to indicate that the Q# program needs to ask the target machine to check the state associated with the inputs to resolve the condition.

**Comment by gillenhaalb** [comment]
@cgranade I agree, a few lines of clarification are certainly in order (it's getting late here, so I'll be taking care of it tomorrow). Changing the `AssertPositive` function name is obviously also necessary, but the question is to what? Perhaps `TestPositive` or `FailIfNotPositive`?

The clarifying discussion would then also serve as a good place to mention the actual `Fact` function in comparison to this to-be-renamed one---namely that this one actually halts the computation, whereas (it appears) `Fact` merely prints a message.

**Comment by cgranade** [comment]
@gillenhaalb: Thanks for taking a look! With respect to `Fact`, it does actually halt if a particular fact is false (that is, `Fact(true, "")` is a tautology while `Fact(false, "")` is a contradiction). From that standpoint, `AssertPositive` could be renamed to `PositivityFact`.

The way I think of it is that a fact is either a tautology or a contradiction as a function only of its value: `true` is always true, irrespective of what the state of any control environment or qubit register is, while an assertion could be true or false depending on state external to the values passed.

The two can even be used well together, by representing global state with values that can then be used with facts. For example:
```Q#
operation AssertQubitsAreAvailable() : Unit {
     Fact(GetQubitsAvailableToUse() > 0, "No qubits were actually available");
}
}
```
That couldn't be a function, since `GetQubitsAvailableToUse` depends on global state of the program and its execution environment (namely, how many qubits are available to use), but it can use that global state to find a value that can be used to formulate a fact.

**Comment by gillenhaalb** [comment]
@cgranade that makes a lot of sense, thanks! I think it may be worth adding that info to the API for [Fact](https://docs.microsoft.com/en-us/qsharp/api/qsharp/microsoft.quantum.diagnostics.fact?view=qsharp-preview) then, because it's hardly clear that it halts the computation (a feature which seems important enough to garner a few sentences of discussion on the Testing and Debugging page). It's only implicit in the description of the input string as "Failure message string to be printed ...".

I'll do an additional PR to the libraries to add the sentence in the function description.

**Comment by gillenhaalb** [comment]
Fixed by #695
